lilypad/llm_fn.py

```diff
@@ -70,26 +85,65 @@ def decorator(
             async def inner_async(*args: _P.args, **kwargs: _P.kwargs) -> _R:
                 arg_types, arg_values = inspect_arguments(fn, *args, **kwargs)
                 version = get_llm_function_version(fn, arg_types, synced)
-                trace_decorator = trace(
-                    project_id=lilypad_client.project_id,
-                    version_id=version.id,
-                    arg_types=arg_types,
-                    arg_values=arg_values,
-                    lexical_closure=version.llm_fn.code,
-                    prompt_template=version.fn_params.prompt_template
-                    if version.fn_params
-                    else "",
-                    version=version.version,
-                )
 
                 if not synced:
+                    trace_decorator = trace(
+                        project_id=lilypad_client.project_id,
+                        version_id=version.id,
+                        arg_types=arg_types,
+                        arg_values=arg_values,
+                        lexical_closure=version.llm_fn.code,
+                        prompt_template=version.fn_params.prompt_template
+                        if version.fn_params
+                        else "",
+                        version=version.version,
+                    )
                     return cast(_R, await trace_decorator(fn)(*args, **kwargs))
 
                 if not version.fn_params:
                     raise ValueError(f"Synced function {fn.__name__} has no params.")
 
+                fn_params = version.fn_params
+
+                @contextmanager
```

Comment:
this looks exactly the same as the one below and doesn't seem to rely on sync vs. async, can we convert this into a utility that returns the context manager instead?

Reply:
done